### PR TITLE
fix(lambda): SAM CLI asset metadata missing from image Functions

### PR DESCRIPTION
**Pull request by torresxb1**
Adds asset metadata to image-type lambda functions. This will allow SAM CLI to support local invocation of image-type lambdas from CDK-synthed templates.

It follows the same design and builds upon https://github.com/aws/aws-cdk/pull/1433

Fixes https://github.com/aws/aws-cdk/issues/14593

Uses some changes from https://github.com/aws/aws-cdk/pull/17293 to enable asset metadata generation in integration tests

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/lib/code.ts [hunk 1]
```diff
@@ -517,28 +517,45 @@ export interface AssetImageCodeProps extends ecr_assets.DockerImageAssetOptions
...
+    if (!this.asset) {
+      this.asset = new ecr_assets.DockerImageAsset(scope, 'AssetImage', {
+        directory: this.directory,
+        ...this.props,
+      });
+      this.asset.repository.grantPull(new iam.ServicePrincipal('lambda.amazonaws.com'));
+    } else if (cdk.Stack.of(this.asset) !== cdk.Stack.of(scope)) {
+      throw new Error(`Asset is already associated with another stack '${cdk.Stack.of(this.asset).stackName}'. ` +
+        'Create a new Code instance for every stack.');
+    }
```
Comment (torresxb1) [inline packages/@aws-cdk/aws-lambda/lib/code.ts:537]: got this by looking at AssetCode above. Let me know if any of this doesn't also work for AssetImageCode

File: packages/@aws-cdk/aws-lambda/test/docker-lambda-handler/Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM 628053151772.dkr.ecr.sa-east-1.amazonaws.com/awslambda/nodejs12.x-runtime-internal:beta
+FROM public.ecr.aws/lambda/nodejs:12 AS stage
```
Comment (torresxb1) [inline packages/@aws-cdk/aws-lambda/test/docker-lambda-handler/Dockerfile:1]: I believe I didn't have access to the previous repo, but this one worked. Added `AS stage` to use docker target in integration test.

File: packages/@aws-cdk/aws-lambda/test/integ.layer-version.lit.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ const awsAccountId = stack.account;
 /// !show
 const layer = new lambda.LayerVersion(stack, 'MyLayer', {
   code: lambda.Code.fromAsset(path.join(__dirname, 'layer-code')),
-  compatibleRuntimes: [lambda.Runtime.NODEJS_10_X],
+  compatibleRuntimes: [lambda.Runtime.NODEJS_14_X],
```
Comment (torresxb1) [inline packages/@aws-cdk/aws-lambda/test/integ.layer-version.lit.ts:15]: got an error when trying to update the integration test, saying that nodejs_10_x was not supported any more

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts [hunk 4]
```diff
@@ -147,6 +147,29 @@ export class DockerImageAsset extends CoreConstruct implements IAsset {
    */
   public readonly assetHash: string;
 
+  /**
+   * The path to the asset, relative to the current Cloud Assembly
+   *
+   * If asset staging is disabled, this will just be the original path.
+   * If asset staging is enabled it will be the staged path.
+   */
+  public readonly assetPath: string;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:156]: Do these all really need to be `public` ? I would be fine if they were `private`, and your feature would still work.

File: packages/@aws-cdk/aws-lambda/lib/code.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-lambda/lib/code.ts:537]: Looks correct, thanks! 👍

File: packages/@aws-cdk/aws-lambda/lib/code.ts [hunk 5]
```diff
@@ -517,28 +517,45 @@ export interface AssetImageCodeProps extends ecr_assets.DockerImageAssetOptions
...
-        imageUri: asset.imageUri,
+        imageUri: this.asset.imageUri,
         entrypoint: this.props.entrypoint,
         cmd: this.props.cmd,
         workingDirectory: this.props.workingDirectory,
       },
     };
   }
+
+  public bindToResource(resource: cdk.CfnResource, options: ResourceBindOptions = { }) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-lambda/lib/code.ts:549]: Was this function already there?

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-lambda/lib/code.ts (hunk 5 above)
Comment (moelasmar) [reply packages/@aws-cdk/aws-lambda/lib/code.ts:549]: yes, this function is already defined in [Code class](https://github.com/aws/aws-cdk/blob/b3eafc2dc61ed69de20196fa08a4df3c29ecc894/packages/@aws-cdk/aws-lambda/lib/code.ts#L152-L154)

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts (hunk 4 above)
Comment (torresxb1) [reply packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts]: updated them

**Review by rix0rrr — APPROVED** [review]

**Review by rix0rrr — APPROVED** [review]

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**